### DISCUSSION TITLE
A database to filter IP traffic

### DISCUSSION THREAD

**Discussion by TomHsiung** [strongswan/strongswan#2422] [Ideas]
Hmmm, I know strongSwan functions at the IP level. It determines the traffic based on both source and target IP addresses (i.e., tunnel mode). I am thinking of a software-based method to, maybe, differentiate those target IPs I would like to bypass the tunnel from those ones that should go into the tunnel.

So, the idea is to introduce a bypass database, which stores those target IP addresses. Before we can bypass traffic via domains, this is the first step.

PS: We have needs to bypass some traffic but strongSwan tunnel are not able to distinguish them without a specific hardware to do so.

Just an idea

**Comment by tobiasbrunner** [comment]
Did you consider marks?

**Comment by TomHsiung** [comment]
>Did you consider marks?

Great. Please let me study the marks function a little bit.

**Comment by TomHsiung** [comment]
iptables rules

> MARK       all  --  10.0.15.21          202.119.189.80       MARK set 0x2

route tables

> 0:	from all lookup local
> 219:	from all fwmark 0x2 lookup 200
> 220:	from all lookup 220
> 32766:	from all lookup main
> 32767:	from all lookup default

custom route table 200

> 202.119.189.80 dev ppp0 scope link

Traceroute shows packets from `10.0.15.21` to `202.119.189.80` still flow into the IPsec tunnel.

**Reply by tobiasbrunner** [reply]
Not sure what you are trying to say here, But did you configure a mark in the config (`mark_out` in [swanctl.conf](https://docs.strongswan.org/docs/5.9/swanctl/swanctlConf.html))? So that only marked packets match the IPsec policy and SA?

**Reply by TomHsiung** [reply]
No. I have not because I don't know this step. And I use the legacy strongSwan file system (e.g., ipsec.conf) rather than the the kind of swanctl.conf. Meanwhile I am googling this configuration, please also let me know the codes for this function. Thank you.

**Reply by TomHsiung** [reply]
By default, we want all traffic flows into the tunnel. We only want to bypass some traffic to specific destinations. Can the mark function of strongSwan achieve this purpose? I think the mark function of it is letting specific traffic flow into the tunnel, instead.

**Reply by tobiasbrunner** [reply]
Then just mark all the traffic except the one you don't want to tunnel (e.g. make the rule not match that traffic, or reset the mark for specific traffic afterwards, or use a custom chain where you return for certain traffic before marking the rest).

**Reply by TomHsiung** [reply]
OK. Thanks. I think I have to configure both gateways at either end of the tunnel. Currently, I use the SNAT rule of a second gateway to modify the source IP addresses of those packets that I want to exclude.

**Comment by TomHsiung** [comment]
Look like the marked packets are still passing to the xfrm lookup step of the iptables system. Therefore, based on the IPsec policy, they are encrypted into the IPsec tunnel. It is necessary to route these marked packets to another gateway. That's a trouble for home users because there is only one ADSL link.
